### ISSUE TITLE
Control's `rect_clip_content` should allow choosing the Axis that is clipped

### ISSUE THREAD

**Issue by Arecher** [godotengine/godot-proposals#2119]
### Describe the project you are working on
Game Project

### Describe the problem or limitation you are having in your project
Whenever I use `Control` nodes that need to clip their content, I find that I only ever really mean to clip the content in either the `vertical axis` or the `horizontal axis`. For example, when I create a scrollable list of items, I only need the list to be clipped in the **vertical** axis: any part of the list that is above or below the rect needs to be clipped. However, parts of the content that are extruding the left or right of the content do not need to be clipped.

![image](https://user-images.githubusercontent.com/10502115/104583116-0b84c880-5661-11eb-82f9-63fc24c442c3.png)

This is something I frequently run into while making iterations to our UI. Small additions need to be made, and for visual appearance or spacing reasons, I designed them to extrude the core rect. Whenever I translate my designs into Godot, I find that it simply isn't possible, due to the rect being too small and cutting off my additions.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
I want the `rect_clip_content` to be split into 2 (or 4: each direction rather than each axis) separate booleans. `rect_clip_vertical` and `rect_clip_horizontal`. That way, I can make sure that vertical scrollable control nodes only ever clip the areas that I want to have clipped.

![image](https://user-images.githubusercontent.com/10502115/104583261-35d68600-5661-11eb-9f6e-2460984b3745.png)
_the way I would like to have the clipped control behave_

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
If `rect_clip_vertical` is `true`, only content above the top red line, and below the bottom red line should be clipped. If `rect_clip_horizontal` is `true`, only content to the left of the left green line, and to the right of the right green line should be clipped.
If both are `true`, any content outside the `rect` should be clipped, as it works now.

![image](https://user-images.githubusercontent.com/10502115/104583575-95349600-5661-11eb-8cb1-72fd0471a39c.png)

### If this enhancement will not be used often, can it be worked around with a few lines of script?
I'm not sure how often other developers run into this, but during the development of our game I have ran into this numerous times already. 

This issue can actually be worked around, but it generally requires a lot of premeditated Control node setup. With multiple `MarginContainers` the core rect can be made bigger than it seems, meaning there is additional space on the left & right, in which nodes can extrude without being clipped. Getting this to work within the allowed space, and without impacting the visual spacing of your UI, can require quite a lot of fiddling. 
Therefore, I still think that the requested feature would be a better, more consistent and easier to understand solution.

### Is there a reason why this should be core and not an add-on in the asset library?
Core: Expands on Core components of Control Nodes

**Comment by KoBeWi** [comment]
Have you tried using ScrollContainer?

**Comment by Arecher** [comment]
> Have you tried using ScrollContainer?

Absolutely. Really dislike working with them though, since they cannot be modified all that well. In terms of visuals (I know you can adjust them with a Theme, but that is very limited.), spacing (you cannot easily add more spacing between the content and the scrollbar, for example) and scripts.
Decided to basically make my own ScrollContainer Scene a while back, since that allows me a lot more freedom and control.

Why though? Do ScrollContainers allow for this? I thought they were a clipped Control node with some additional nodes/functionality. I presumed they worked the exact same way.

**Comment by KoBeWi** [comment]
If you disable horizontal scrolling, they only clip vertically. And you can control spacing with MarginContainer.

What functionality do you think they lack in theming and scripting?

**Comment by Arecher** [comment]
> If you disable horizontal scrolling, they only clip vertically. And you can control spacing with MarginContainer.

I just tried that, but it still clips horizontally as well. (also when running the project)
![image](https://user-images.githubusercontent.com/10502115/104589717-93bb9b80-566a-11eb-9d0a-078771d8e566.png)


> What functionality do you think they lack in theming and scripting?

Now that's a can of worms. I never made any requests regarding Scollcontainers, because I gave up on them a long time ago. I'll see if I can make some time to mess around with them again sometime, and create requests for all the limitations I find.

**Comment by KoBeWi** [comment]
> I just tried that, but it still clips horizontally as well. (also when running the project)

I wasn't precise - it's not like the container won't clip, it will fit to its content. So if you want your UI to display properly, everything needs to be correctly placed inside containers. From what I see on your scene, the "new" label is a child of plain Control node, which makes it not fit properly.

If you want to achieve what is on the left, then yeah, it's impossible. You can't make something that will appear on top of scrollbar. Your labels either have to fit inside the panels or you need to add a margin on the right.

Your request is a bit problematic though. Clipping uses rectangles. If you want to clip on one axis, you need a rectangle infinite on the other axis. Not sure how could that be implemented. Maybe we could just add/expose a method for setting a custom clipping rectangle and you would be supposed to provide the rect that fits your needs.
